From 5186aa8c236e13af5619cb7ca1875f79a3b759dd Mon Sep 17 00:00:00 2001
From: Ray Donnelly <mingw.android@gmail.com>
Date: Fri, 27 Apr 2018 13:02:40 -0700
Subject: [PATCH 01/14] intel_mkl version

---
 numpy/__init__.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/numpy/__init__.py b/numpy/__init__.py
index db99294..99c954f 100644
--- a/numpy/__init__.py
+++ b/numpy/__init__.py
@@ -197,3 +197,5 @@ else:
     # but do not use them, we define them here for backward compatibility.
     oldnumeric = 'removed'
     numarray = 'removed'
+
+__mkl_version__ = "__MKL_VERSION__"
-- 
1.9.5 (Apple Git-50.3)

