From 60911202a6886263e6975161934116a028e3fa98 Mon Sep 17 00:00:00 2001
From: Michael Sarahan <msarahan@gmail.com>
Date: Thu, 26 Jul 2018 11:30:25 -0500
Subject: [PATCH 07/14] define mkl_version in __init__.py

---
 numpy/__init__.py | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/numpy/__init__.py b/numpy/__init__.py
index d250ed5ac..2916694ab 100644
--- a/numpy/__init__.py
+++ b/numpy/__init__.py
@@ -196,6 +196,8 @@ else:
     del PytestTester
 
 
+    __mkl_version__ = "__MKL_VERSION__"
+
     def _sanity_check():
         """
         Quick sanity checks for common bugs caused by environment.
-- 
2.17.1

