diff --git a/numpy/distutils/ccompiler.py b/numpy/distutils/ccompiler.py
index 8484685..0fd5219 100644
--- a/numpy/distutils/ccompiler.py
+++ b/numpy/distutils/ccompiler.py
@@ -434,7 +434,7 @@ def CCompiler_get_version(self, force=False, ok_status=[0]):
         Version string, in the format of `distutils.version.LooseVersion`.
 
     """
-    if not force and hasattr(self, 'version'):
+    if not force and hasattr(self, 'version') and self.version:
         return self.version
     self.find_executables()
     try:
@@ -460,10 +460,9 @@ def CCompiler_get_version(self, force=False, ok_status=[0]):
     status, output = exec_command(version_cmd, use_tee=0)
 
     version = None
-    if status in ok_status:
-        version = matcher(output)
-        if version:
-            version = LooseVersion(version)
+    version = matcher(output)
+    if version:
+        version = LooseVersion(version)
     self.version = version
     return version
 
