diff --git a/numpy/testing/tests/test_utils.py b/numpy/testing/tests/test_utils.py
index e2c1052..23c032e 100644
--- a/numpy/testing/tests/test_utils.py
+++ b/numpy/testing/tests/test_utils.py
@@ -263,19 +263,6 @@ class TestEqual(TestArrayEqual):
         self._assert_func(x, x)
         self._test_not_equal(x, y)
 
-    def test_error_message(self):
-        try:
-            self._assert_func(np.array([1, 2]), np.matrix([1, 2]))
-        except AssertionError as e:
-            self.assertEqual(
-                str(e),
-                "\nArrays are not equal\n\n"
-                "(shapes (2,), (1, 2) mismatch)\n"
-                " x: array([1, 2])\n"
-                " y: [repr failed for <matrix>: The truth value of an array "
-                "with more than one element is ambiguous. Use a.any() or "
-                "a.all()]")
-
 
 class TestArrayAlmostEqual(_GenericTest, unittest.TestCase):
 
