diff --git a/numpy/core/tests/test_memmap.py b/numpy/core/tests/test_memmap.py
index c0c3522..d69f2b5 100644
--- a/numpy/core/tests/test_memmap.py
+++ b/numpy/core/tests/test_memmap.py
@@ -27,7 +27,10 @@ class TestMemmap(TestCase):
 
     def tearDown(self):
         self.tmpfp.close()
-        shutil.rmtree(self.tempdir)
+        try:
+            shutil.rmtree(self.tempdir)
+        except:
+            pass
 
     def test_roundtrip(self):
         # Write data to file
@@ -83,9 +86,9 @@ class TestMemmap(TestCase):
                        shape=self.shape)
         abspath = os.path.realpath(os.path.abspath(tmpname))
         fp[:] = self.data[:]
-        self.assertEqual(abspath, str(fp.filename.resolve()))
+        self.assertEqual(abspath.lower(), str(fp.filename.resolve()).lower())
         b = fp[:1]
-        self.assertEqual(abspath, str(b.filename.resolve()))
+        self.assertEqual(abspath.lower(), str(b.filename.resolve()).lower())
         del b
         del fp
 
